codefeedr-plugins/codefeedr-travis/src/main/scala/org/codefeedr/plugins/travis/util/TravisBuildCollector.scala

```diff
@@ -0,0 +1,158 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package org.codefeedr.plugins.travis.util
+
+import java.time.{LocalDateTime, ZoneId}
+import org.codefeedr.plugins.travis.TravisProtocol.{TravisBuild, TravisBuilds}
+import org.codefeedr.plugins.travis.util.TravisExceptions._
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.{blocking, Future}
+
+/**
+  * Class that keeps querying Travis about a certain push event until it is found.
+  * @param repoOwner Owner of the repo of the push event
+  * @param repoName Name of the repo of the push event
+  * @param branchName Name of the branch of the push event
+  * @param pushCommitSha Sha of the commit of the push event
+  * @param pushDate Date of the push event
+  * @param travis TravisService used to query Travis
+  * @param pollingInterval Interval at which builds are queried
+  * @param timeoutSeconds How long it will wait for the build to be created
+  */
+class TravisBuildCollector(repoOwner: String,
+                           repoName: String,
+                           branchName: String,
+                           pushCommitSha: String,
+                           pushDate: LocalDateTime,
+                           travis: TravisService,
+                           pollingInterval: Int = 30000,
+                           timeoutSeconds: Int = 60) {
+
+  private var minimumStartDate: LocalDateTime = pushDate
+  private var build: Option[TravisBuild] = None
+
+
+  /**
+    * Keeps requesting the build until it is finished
+    * @return A finished Travis build
+    */
+  def requestFinishedBuild(): Future[TravisBuild] = Future {
+    while (!isReady) {
+
+      build = requestBuild
+      checkIfBuildShouldBeKnownAlready()
+
+      if (!isReady) {
+        blocking {
+          Thread.sleep(pollingInterval)
+        }
+      }
+    }
+    build.get
+  }
+
+  def checkIfBuildShouldBeKnownAlready(): Unit = {
+    val waitUntil = pushDate.plusSeconds(100 * timeoutSeconds)
+    if (build.isEmpty && waitUntil.isBefore(LocalDateTime.now(ZoneId.of("GMT")))) {
+      throw BuildNotFoundForTooLongException("Waited " + timeoutSeconds + " seconds for build, but still not found" +
```

Comment:
Does this crash the whole Flink job? If so, is that the behavior we want?

Reply:
No this does not crash the whole flink job